### PR TITLE
BUGFIX: Node fallbacks not properly checking for variants

### PR DESCRIPTION
**Pull request by kitsunet**
This adds a check for better matching node variants to path traversal
queries of the NodeDataRepository to prevent fallbacks showing up
even if a better matching variant exists.

NEOS-1841 #resolve

### CODE REVIEW

**Comment by kitsunet** [comment]
@kdambekalns

**Comment by kitsunet** [comment]
StyleCI is unrelated :(

**Comment by kitsunet** [comment]
ok, waiting for travis, I *think* this version finally works as intended... What a ride...

**Comment by kitsunet** [comment]
@kdambekalns I think the unit test failures are due to the changed methods which makes a readjustment of mocks necessary. Maybe you could fix that and then it should be good to go.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 1]
```diff
@@ -1476,4 +1530,69 @@ protected function addIdentifierConstraintToQueryBuilder(QueryBuilder $queryBuil
...
+    /**
+     * @param array $nodeDataObjects
+     * @param Workspace $workspace
+     * @param array $dimensions
+     * @return array
+     */
+    protected function filterNodeDataByBestMatchInContext(array $nodeDataObjects, Workspace $workspace, array $dimensions)
+    {
+        $workspaces = $this->collectWorkspaceAndAllBaseWorkspaces($workspace);
+        $notPersistedNodes = [];
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: nonPersisted?

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 2]
```diff
@@ -1476,4 +1530,69 @@ protected function addIdentifierConstraintToQueryBuilder(QueryBuilder $queryBuil
...
+        });
+    }
+
+    /**
+     * Returns an array that contains the given workspace and all base (parent) workspaces of it.
+     *
+     * @param Workspace $workspace
+     * @return array
+     */
+    protected function collectWorkspaceAndAllBaseWorkspaces(Workspace $workspace)
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: this method can be used in lines 534, 573, 626, 930, 998 & 1384

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 3]
```diff
@@ -170,29 +172,76 @@ public function remove($object)
...
+
+    /**
+     * @param string $path
+     * @param Workspace $workspace
+     * @param array|null $dimensions
+     * @param boolean $onlyShadowNodes
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    protected function prepareFindNodesByPath($path, Workspace $workspace, array $dimensions = null, $onlyShadowNodes = false)
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: method name seems a bit off.. not sure about a better name though

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php (hunk 2 above)
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: Yes I know, I didn't want to make this into a big refactoring as it's a bugfix in lower versions. I think I didn something similar for 2.2 or master already

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 4]
```diff
@@ -170,29 +172,81 @@ public function remove($object)
...
+     * removed nodes must be done on the result.
+     *
+     * @param string $path
+     * @param Workspace $workspace
+     * @param array|null $dimensions
+     * @param boolean $onlyShadowNodes
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    protected function findRawNodesByPath($path, Workspace $workspace, array $dimensions = null, $onlyShadowNodes = false)
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: Nice name 👍

**Comment by kitsunet** [comment]
👍

**Comment by kitsunet** [comment]
IMHO merge ready, unless you want to do the refactoring for the workspace collection thingy... From what I see that is indeed still missing in master, so could be done here already....

**Comment by aertmann** [comment]
yep looks goods, well done 👍

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php (hunk 2 above)
Comment (kdambekalns) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: I did that in my last commit here, so…

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php (hunk 4 above)
Comment (kdambekalns) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: Thanks! 😊

**Comment by kdambekalns** [comment]
So, this not only passes the tests now but indeed fixes the issue it was supposed to fix in our customer project. 🎱

**Pull request merged**